server/quota/quota.go

```diff
@@ -50,215 +49,102 @@ type State struct {
-	collSize, err := tenant.CollectionSize(ctx, db, coll)
-	if err != nil {
-		ulog.E(err)
-	}
-	return collSize
+	return nil
 }
 
-func (m *Manager) getTenant(ctx context.Context, namespace string) *metadata.Tenant {
-	tenant, err := m.tenantMgr.GetTenant(ctx, namespace, m.txMgr)
-	ulog.E(err)
-	return tenant
+func (m *Manager) cleanup() error {
+	// TODO: call cleanup of all quota limiters
+	return nil
 }
 
-func (m *Manager) getTenantSize(ctx context.Context, namespace string) int64 {
-	tenant := m.getTenant(ctx, namespace)
-	if tenant == nil {
-		return 0
-	}
-	size, err := tenant.Size(ctx)
-	ulog.E(err)
-	return size
+func Cleanup() {
+	_ = mgr.cleanup()
 }
 
-func (m *Manager) updateMetricsForNamespace(ctx context.Context, namespace string) {
-	if m.txMgr == nil {
-		return
-	}
-
-	tenant := m.getTenant(ctx, namespace)
-	if tenant == nil {
-		return
+func unitSize(isWrite bool) int64 {
+	if isWrite {
+		return int64(config.WriteUnitSize)
 	}
 
-	for _, dbName := range tenant.ListDatabases(ctx) {
-		db, err := tenant.GetDatabase(ctx, dbName)
-		if err != nil {
-			ulog.E(err)
-			return
-		}
-		metrics.UpdateDbSizeMetrics(namespace, dbName, getDbSize(ctx, tenant, db))
-		for _, coll := range db.ListCollection() {
-			metrics.UpdateCollectionSizeMetrics(namespace, dbName, coll.Name, getCollSize(ctx, tenant, db, coll))
-		}
-	}
-
-	metrics.UpdateNameSpaceSizeMetrics(namespace, m.getTenantSize(ctx, namespace))
-}
-
-func (m *Manager) updateAllMetrics(ctx context.Context) {
-	for _, namespace := range m.tenantMgr.GetNamespaceNames() {
-		m.updateMetricsForNamespace(ctx, namespace)
-	}
+	return int64(config.ReadUnitSize)
 }
 
-func (m *Manager) updateTenantSize(ctx context.Context, namespace string) {
-	s := m.getState(namespace)
-	s.SizeLock.Lock()
-	defer s.SizeLock.Unlock()
+func toUnits(size int, isWrite bool) int {
+	us := unitSize(isWrite)
 
-	currentTimeStamp := time.Now().Unix()
-	s.SizeUpdateAt.Store(currentTimeStamp)
-
-	dsz := m.getTenantSize(ctx, namespace)
-	s.Size.Store(dsz)
+	// + (unitSize - 1) guarantees that fractional part is counted as one unit
+	return int((int64(size) + us - 1) / us)
 }
 
-func (m *Manager) updateTenantMetrics(ctx context.Context, namespace string) {
-	s := m.getState(namespace)
-	sz := s.Size.Load()
-	currentTimeStamp := time.Now().Unix()
-
-	if currentTimeStamp >= s.TenantSizeUpdateAt.Load()+m.cfg.TenantSizeRefreshInterval {
-		s.TenantSizeLock.Lock()
-		defer s.TenantSizeLock.Unlock()
-
-		s.TenantSizeUpdateAt.Store(currentTimeStamp)
-		metrics.UpdateNameSpaceSizeMetrics(namespace, sz)
-		m.updateTenantSize(ctx, namespace)
+// Allow checks read, write rates and storage size limits for the namespace
+// and returns error if at least one of them is exceeded.
+func Allow(ctx context.Context, namespace string, size int, isWrite bool) error {
+	for _, q := range mgr.quota {
+		if err := q.Allow(ctx, namespace, size, isWrite); err != nil {
```

Comment:
Here best to have metrics if we are not allowing and which rate-limiter is kicking in.